File: my-solutions/0054-spiral-matrix/solution.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        int row = matrix.size(), col = matrix[0].size();
        int total = row * col, count = 0;
        int start_row = 0, start_col = 0, end_row = row - 1, end_col = col - 1;

        while (count < total) {
            for (int i = start_col; count < total && i <= end_col; i++) {
                ans.push_back(matrix[start_row][i]);
                count++;
            }
            start_row++;

            for (int i = start_row; count < total && i <= end_row; i++) {
                ans.push_back(matrix[i][end_col]);
                count++;
            }
            end_col--;

            for (int i = end_col; count < total && i >= start_col; i--) {
                ans.push_back(matrix[end_row][i]);
                count++;
            }
            end_row--;

            for (int i = end_row; count < total && i >= start_row; i--) {
                ans.push_back(matrix[i][start_col]);
                count++;
            }
            start_col++;
        }

        return ans;
    }
};

File: my-solutions/0054-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        visited = []
        for i in range(m):
            row = [False] * n
            visited.append(row)
        
        spiral = []
        dr, dc = 0, 1
        r, c = 0, 0

        while len(spiral) < m * n:
            spiral.append(matrix[r][c])
            visited[r][c] = True

            if dr == 0 and dc == 1 and (c + dc >= n or visited[r][c + dc]):
                dr, dc = 1, 0
            elif dr == 1 and dc == 0 and (r + dr >= m or visited[r + dr][c]):
                dr, dc = 0, -1
            elif dr == 0 and dc == -1 and (c + dc < 0 or visited[r][c + dc]):
                dr, dc = -1, 0
            elif dr == -1 and dc == 0 and (r + dr < 0 or visited[r + dr][c]):
                dr, dc = 0, 1
            
            r += dr
            c += dc
        
        return spiral
